File: backend/src/iaas_sim/application/get_operation.py
from iaas_sim.application.operation import (
    BackendOperationFailed,
    BackendOperationPort,
    BackendOperationRunning,
    BackendOperationStatus,
    BackendOperationSucceeded,
    OperationNotFound,
    OperationPersistenceFailure,
    OperationPollingFailure,
    OperationStorePort,
    StoredOperation,
)
from iaas_sim.domain.entity.operation import (
    Failed,
    Operation,
    OperationFailure,
    OperationId,
    Succeeded,
    is_terminal,
)
from iaas_sim.result import Ok, Result, and_then, map_error

BACKEND_OPERATION_FAILURE_REASON = "Backend operation failed"
type GetOperationError = OperationNotFound | OperationPersistenceFailure | OperationPollingFailure


def _as_get_operation_error(error: GetOperationError) -> GetOperationError:
    return error


def get_operation(
    store: OperationStorePort, backend: BackendOperationPort, operation_id: OperationId
) -> Result[Operation, GetOperationError]:
    def poll(stored: StoredOperation) -> Result[Operation, GetOperationError]:
        if is_terminal(stored.operation.status):
            return Ok(stored.operation)

        def reconcile(status: BackendOperationStatus) -> Result[Operation, GetOperationError]:
            match status:
                case BackendOperationRunning():
                    return Ok(stored.operation)
                case BackendOperationSucceeded():
                    terminal = Operation(
                        stored.operation.id,
                        stored.operation.target,
                        stored.operation.action,
                        Succeeded(),
                    )
                case BackendOperationFailed():
                    terminal = Operation(
                        stored.operation.id,
                        stored.operation.target,
                        stored.operation.action,
                        Failed(OperationFailure(BACKEND_OPERATION_FAILURE_REASON)),
                    )
            return map_error(store.complete(terminal), _as_get_operation_error)

        return and_then(backend.get_operation_status(stored.backend_ref), reconcile)

    return and_then(store.get(operation_id), poll)

File: backend-kotlin/src/main/kotlin/dev/iaassim/application/GetOperation.kt
package dev.iaassim.application

import dev.iaassim.domain.entity.operation.Failed
import dev.iaassim.domain.entity.operation.Operation
import dev.iaassim.domain.entity.operation.OperationFailure
import dev.iaassim.domain.entity.operation.OperationId
import dev.iaassim.domain.entity.operation.Succeeded
import dev.iaassim.domain.entity.operation.isTerminal
import dev.iaassim.result.Err
import dev.iaassim.result.Ok
import dev.iaassim.result.Outcome

fun getOperation(store: OperationStorePort, backend: BackendOperationPort, operationId: OperationId):
    Outcome<Operation, GetOperationError> {
    val stored = when (val result = store.get(operationId)) {
        is Ok -> result.value
        is Err -> return when (val error = result.error) {
            is OperationNotFound -> Err(error)
            is OperationPersistenceFailure -> Err(error)
        }
    }
    if (isTerminal(stored.operation.status)) return Ok(stored.operation)
    val terminal = when (val result = backend.getOperationStatus(stored.backendRef)) {
        is Err -> return Err(result.error)
        is Ok -> when (result.value) {
            BackendOperationRunning -> return Ok(stored.operation)
            BackendOperationSucceeded -> stored.operation.copy(status = Succeeded)
            is BackendOperationFailed -> stored.operation.copy(status = Failed(OperationFailure("Backend operation failed")))
        }
    }
    return when (val result = store.complete(terminal)) {
        is Ok -> result
        is Err -> when (val error = result.error) {
            is OperationNotFound -> Err(error)
            is OperationPersistenceFailure -> Err(error)
        }
    }
}
